File: src/models/ai_trace.py
"""AI Query Trace model for end-to-end AI journey tracing."""

from sqlalchemy import (
    Column, Integer, BigInteger, Numeric, String, DateTime, Text,
    Boolean, ForeignKey, JSON, func,
)
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import relationship

from src.config.database import Base


class AIQueryTrace(Base):
    """Represents a single span in an AI query trace tree.

    Span types:
    - query: Root span for the entire AI query lifecycle
    - llm_call: Individual LLM API call (one per iteration)
    - tool_execution: Single tool call with input/output/timing
    - synthesis: Final synthesis/response generation step
    """
    __tablename__ = "ai_query_traces"

    id = Column(Integer, primary_key=True, autoincrement=True)
    trace_id = Column(PG_UUID(as_uuid=True), nullable=False, index=True)
    parent_span_id = Column(Integer, ForeignKey("ai_query_traces.id", ondelete="CASCADE"), nullable=True)
    session_id = Column(Integer, ForeignKey("ai_sessions.id", ondelete="CASCADE"), nullable=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=True, index=True)

    # Span identification
    span_type = Column(String(30), nullable=False, index=True)
    span_name = Column(String(255), nullable=True)
    iteration = Column(Integer, default=0)

    # Timing
    start_time = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    end_time = Column(DateTime(timezone=True), nullable=True)
    duration_ms = Column(Integer, nullable=True)

    # LLM call fields
    model = Column(String(100), nullable=True)
    provider = Column(String(30), nullable=True)
    input_tokens = Column(BigInteger, nullable=True)
    output_tokens = Column(BigInteger, nullable=True)
    cost_usd = Column(Numeric(12, 8), nullable=True)
    thinking_tokens = Column(BigInteger, nullable=True)

    # Tool execution fields
    tool_name = Column(String(255), nullable=True)
    tool_input = Column(JSON, nullable=True)
    tool_output_summary = Column(Text, nullable=True)
    tool_success = Column(Boolean, nullable=True)
    tool_platform = Column(String(30), nullable=True)
    tool_error = Column(Text, nullable=True)

    # Network timing
    dns_ms = Column(Integer, nullable=True)
    tcp_connect_ms = Column(Integer, nullable=True)
    tls_ms = Column(Integer, nullable=True)
    ttfb_ms = Column(Integer, nullable=True)

    # Network path info (captured from instrumented HTTP transport)
    server_ip = Column(String(45), nullable=True)       # Resolved IP (v4 or v6)
    server_port = Column(Integer, nullable=True)
    tls_version = Column(String(20), nullable=True)      # e.g., "TLSv1.3"
    http_version = Column(String(10), nullable=True)      # e.g., "HTTP/2"
    network_path = Column(JSON, nullable=True)            # Full path hops from traceroute/TE

    # Status
    status = Column(String(20), default="running")
    error_message = Column(Text, nullable=True)
    trace_metadata = Column("metadata", JSON, default=dict)

    # Self-referential relationship
    parent = relationship("AIQueryTrace", remote_side=[id], backref="children")

File: web-ui/src/types/ai-trace.ts
export type SpanType = 'query' | 'llm_call' | 'tool_execution' | 'synthesis';
export type SpanStatus = 'running' | 'success' | 'error' | 'timeout';

export interface AITraceSpan {
  id: number;
  trace_id: string;
  parent_span_id: number | null;
  span_type: SpanType;
  span_name: string | null;
  iteration: number;
  start_time: string;
  end_time: string | null;
  duration_ms: number | null;
  // LLM fields
  model: string | null;
  provider: string | null;
  input_tokens: number | null;
  output_tokens: number | null;
  cost_usd: number | null;
  // Tool fields
  tool_name: string | null;
  tool_input: Record<string, unknown> | null;
  tool_output_summary: string | null;
  tool_success: boolean | null;
  tool_platform: string | null;
  tool_error: string | null;
  // Network timing
  dns_ms: number | null;
  tcp_connect_ms: number | null;
  tls_ms: number | null;
  ttfb_ms: number | null;
  // Network path info
  server_ip: string | null;
  server_port: number | null;
  tls_version: string | null;
  http_version: string | null;
  network_path: NetworkHop[] | null;
  // Status
  status: SpanStatus;
  error_message: string | null;
  metadata?: Record<string, unknown>;
  children?: AITraceSpan[];
  // Journey enrichment (from /journey endpoint)
  baseline?: { tcpMs: number | null; tlsMs: number | null; ttfbMs: number | null; durationMs: number | null; isValid: boolean };
  anomalies?: { tcpSlow: boolean; tlsSlow: boolean; ttfbSlow: boolean; durationSlow: boolean };
  // TE enrichment (from /journey endpoint with full TE data)
  te_enrichment?: {
    path_hops: { ip: string; prefix: string; delay: number; loss: number; network: string; rdns: string; asNumber?: string; location?: string }[];
    network_metrics: { loss: number; latency: number; jitter: number; bandwidth?: number } | null;
    bgp_routes: { prefix: string; asPath: number[]; reachability: number; updates: number; monitor: string }[];
    http_timing: { dnsTime: number; connectTime: number; sslTime: number; waitTime: number; receiveTime: number; responseTime: number; wireSize: number; throughput?: number } | null;
    test_id: number | null;
    test_type: string | null;
    agent_name: string | null;
  };
  cost_impact?: {
    cost_usd: number;
    baseline_latency_ms: number | null;
    actual_latency_ms: number | null;
    excess_latency_ms: number;
    wasted_compute_usd: number;
  };
}

export interface AITraceSummary {
  trace_id: string;
  query: string;
  start_time: string;
  duration_ms: number | null;
  status: SpanStatus;
  tool_count: number;
  cost_usd: number;
  network_latency_ms?: number | null;
  provider?: string;
  model?: string;
}

export interface NetworkHop {
  ip: string;
  prefix?: string;
  delay: number;
  loss?: number;
  network?: string;    // e.g., "AS15169 Google LLC"
  rdns?: string;       // Reverse DNS hostname
}

export interface WaterfallBar {
  span_id: number;
  span_type: SpanType;
  span_name: string;
  offset_ms: number;
  duration_ms: number;
  depth: number;
  status: SpanStatus;
  tool_name?: string;
  tool_platform?: string;
  model?: string;
  tokens?: { input: number; output: number };
  cost_usd?: number;
  network_timing?: { dns_ms?: number; tcp_ms?: number; tls_ms?: number; ttfb_ms?: number };
  server_ip?: string;
  server_port?: number;
  tls_version?: string;
  http_version?: string;
  network_path?: NetworkHop[];
}

export interface AITraceDetail {
  trace_id: string;
  root_span: AITraceSpan;
  total_tokens: number;
  total_cost: number;
  tool_count: number;
  span_count: number;
  cost_summary?: {
    total_cost_usd: number;
    total_network_tax_ms: number;
    total_wasted_usd: number;
    network_tax_pct: number;
  };
  te_monitoring?: Record<string, boolean>;
}

export const SPAN_COLORS: Record<SpanType, { bg: string; hex: string }> = {
  query: { bg: 'bg-slate-500', hex: '#64748b' },
  llm_call: { bg: 'bg-blue-500', hex: '#3b82f6' },
  tool_execution: { bg: 'bg-emerald-500', hex: '#10b981' },
  synthesis: { bg: 'bg-purple-500', hex: '#a855f7' },
};

export const PLATFORM_COLORS: Record<string, string> = {
  meraki: '#00bceb',
  catalyst: '#049fd9',
  thousandeyes: '#ff6b35',
  splunk: '#65a637',
  knowledge: '#9333ea',
  canvas: '#6366f1',
};
